src/MediatR/Pipeline/RequestExceptionProcessorBehavior.cs

```diff
@@ -0,0 +1,77 @@
+using System.Reflection;
+
+namespace MediatR.Pipeline
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Collections.Concurrent;
+    using System.Linq;
+    using System.Threading;
+    using System.Threading.Tasks;
+
+    /// <summary>
+    /// Behavior for executing all <see cref="IRequestExceptionHandler{TRequest,TResponse,TException}"/> or <see cref="IRequestExceptionHandler{TRequest,TResponse}"/> instances after an exception is thrown by the request handler
+    /// </summary>
+    /// <typeparam name="TRequest">Request type</typeparam>
+    /// <typeparam name="TResponse">Response type</typeparam>
+    public class RequestExceptionProcessorBehavior<TRequest, TResponse> : IPipelineBehavior<TRequest, TResponse>
+    {
+        private delegate Task HandleExceptionMethod(TRequest request, Exception exception, RequestExceptionHandlerState<TResponse> state, CancellationToken cancellationToken);
+        private readonly ServiceFactory _serviceFactory;
+        private readonly IEnumerable<IRequestExceptionHandler<TRequest, TResponse>> _exceptionHandlers;
+        private readonly ConcurrentDictionary<Type, IEnumerable<HandleExceptionMethod>> _exceptionHandlersCache = new ConcurrentDictionary<Type, IEnumerable<HandleExceptionMethod>>();
```

Comment:
The more I think about it - this can't be static. Is there a way that you can statically cache the execution plan?

Or maybe we just remove the cache for now as a first pass.